### PR TITLE
[serverless] Protect ExecutionContext with mutex

### PR DESCRIPTION
**Pull request by nhinsch**
### What does this PR do?

Protects the ExecutionContext (current invocation state) with a mutex.

### Motivation

Previously, multiple goroutines read and modified the ExecutionContext concurrently in an unsafe way, creating a data race.

### Possible Drawbacks / Trade-offs

Adding a mutex increases the complexity of the code and could create contention between goroutines.

### Describe how to test/QA your changes

Integration tests and unit tests cover these changes.

### Reviewer's Checklist
- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] Changed code has automated tests for its functionality.
- [x] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.

### CODE REVIEW

**Comment by nhinsch** [comment]
Note: Current integration test failures are expected, I will re-run the tests after https://github.com/DataDog/datadog-agent/pull/11091 is merged.

File: pkg/serverless/logs/logs.go [hunk 1]
```diff
@@ -27,27 +28,18 @@ type Tags struct {
...
 	Demux                  aggregator.Demultiplexer
 	ExtraTags              *Tags
-	ExecutionContext       *ExecutionContext
 	LogsEnabled            bool
 	EnhancedMetricsEnabled bool
 	// HandleRuntimeDone is the function to be called when a platform.runtimeDone log message is received
-	HandleRuntimeDone func()
+	HandleRuntimeDone                  func()
+	GetExecutionContext                func() executioncontext.ExecutionContext
+	UpdateExecutionContextFromStartLog func(string, time.Time)
```
Comment (maxday) [inline pkg/serverless/logs/logs.go:42]: I would rather not expose this function from here, but more as a method in `executioncontext.ExecutionContext`
Something like : 
`func (ec *ExecutionContext) UpdateExecutionContextFromStartLog`
What do you think?

Comment (maxday) [inline pkg/serverless/logs/logs.go:42]: I would rather not expose this function from here, but more as a method in `executioncontext.ExecutionContext`
Something like : 
`func (ec *ExecutionContext) UpdateExecutionContextFromStartLog`
What do you think?

File: pkg/serverless/daemon/executioncontext.go [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+// GetExecutionContext gets the current execution context
+func (d *Daemon) GetExecutionContext() executioncontext.ExecutionContext {
+	d.executionContextMutex.Lock()
+	defer d.executionContextMutex.Unlock()
+	return *d.executionContext
+}
+
+// SetExecutionContextFromInvocation sets the execution context based on an invocation
+func (d *Daemon) SetExecutionContextFromInvocation(arn string, requestID string) {
```
Comment (maxday) [inline pkg/serverless/daemon/executioncontext.go:25]: since those methods are tied to `ExecutionContext`, maybe we should attach then to a `ExecutionContext` and not the `Daemon`?

File: pkg/serverless/daemon/executioncontext.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+	if err != nil {
+		return err
+	}
+	d.executionContext.ARN = restoredExecutionContext.ARN
+	d.executionContext.LastRequestID = restoredExecutionContext.LastRequestID
+	d.executionContext.LastLogRequestID = restoredExecutionContext.LastLogRequestID
+	d.executionContext.ColdstartRequestID = restoredExecutionContext.ColdstartRequestID
+	d.executionContext.StartTime = restoredExecutionContext.StartTime
+	return nil
+}
```
Comment (maxday) [inline pkg/serverless/daemon/executioncontext.go:80]: `UpdateExecutionContextFromStartLog`, `RestoreCurrentStateFromFile` and `SaveCurrentExecutionContext` are not covered by unit tests

File: pkg/serverless/daemon/daemon.go [hunk 4]
```diff
@@ -345,7 +342,8 @@ func (d *Daemon) WaitForDaemon() {
 // ComputeGlobalTags extracts tags from the ARN, merges them with any user-defined tags and adds them to traces, logs and metrics
 func (d *Daemon) ComputeGlobalTags(configTags []string) {
 	if len(d.ExtraTags.Tags) == 0 {
-		tagMap := tags.BuildTagMap(d.ExecutionContext.ARN, configTags)
+		ecs := d.ExecutionContext.GetCurrentState()
+		tagMap := tags.BuildTagMap(ecs.ARN, configTags)
```
Comment (maxday) [inline pkg/serverless/daemon/daemon.go:346]: I think we need a `getARN` protected by a lock here because of the following scenario:

1 - ecs := d.ExecutionContext.GetCurrentState() is executed, lock has been acquired and released
2 - another go routine is triggered, modifying ARN
3 - tagMap := tags.BuildTagMap(ecs.ARN, configTags) is executed, ecs.ARN might be in a modification state

File: pkg/serverless/daemon/routes.go [hunk 5]
```diff
@@ -70,10 +70,11 @@ type EndInvocation struct {
 func (e *EndInvocation) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	log.Debug("Hit on the serverless.EndInvocation route.")
 	endTime := time.Now()
+	ecs := e.daemon.ExecutionContext.GetCurrentState()
 	var endDetails = invocationlifecycle.InvocationEndDetails{
 		EndTime:   endTime,
 		IsError:   r.Header.Get(invocationlifecycle.InvocationErrorHeader) == "true",
-		RequestID: e.daemon.ExecutionContext.LastRequestID,
+		RequestID: ecs.LastRequestID,
```
Comment (maxday) [inline pkg/serverless/daemon/routes.go:77]: same comment with a potential data race here?

File: pkg/serverless/daemon/daemon.go (hunk 4 above)
Comment (nhinsch) [reply pkg/serverless/daemon/daemon.go:346]: I'm not sure I follow. Once we call `GetCurrentState()` we have a snapshot of the execution context (`ecs`). If the underlying execution context is modified, the snapshot will be out of date, but there won't be any contention between this goroutine and any other. Does that make sense?

Comment (nhinsch) [reply pkg/serverless/daemon/daemon.go:346]: In other words, `ecs` is a copy, and only this goroutine has a reference to it.

Comment (maxday) [reply pkg/serverless/daemon/daemon.go:346]: yes you're right, ARN is not written again here

**Review by maxday — APPROVED** [review]
LGTM!

File: docs/serverless/README.md [hunk 6]
```diff
@@ -1,52 +0,0 @@
-# Serverless Agent
```
Comment (mx-psi) [inline docs/serverless/README.md:1]: How is this change related to the rest of the PR? Also, `docs/serverless` if kept should be owned by serverless, not Agent Platform

Comment (nhinsch) [reply docs/serverless/README.md:1]: Yeah I'm going to remove this change, it's not related to the rest of the PR. Sorry for the review spam.
